src/sdk/pynni/nni/bohb_advisor/config_generator.py

```diff
@@ -0,0 +1,326 @@
+
+                    try:
+                        sample = ConfigSpace.util.deactivate_inactive_hyperparameters(
+                            configuration_space=self.configspace,
+                            configuration=sample)
+                        info_dict['model_based_pick'] = True
+
+                    except Exception as e:
+                        logger.warning(("="*50 + "\n")*3 +\
+                                "Error converting configuration:\n%s"%sample+\
+                                "\n here is a traceback:" +\
+                                traceback.format_exc())
+                        raise(e)
+            except:
+                logger.warning("Sampling based optimization with %i samples failed\n %s \n \
+                Using random configuration"%(self.num_samples, traceback.format_exc()))
+                sample = self.configspace.sample_configuration()
+                info_dict['model_based_pick'] = False
+
+        try:
+            sample = ConfigSpace.util.deactivate_inactive_hyperparameters(
+                configuration_space=self.configspace,
+                configuration=sample.get_dictionary()
+            ).get_dictionary()
+        except Exception as e:
+            logger.warning("Error (%s) converting configuration: %s -> "
+                        "using random configuration!", e, sample)
+            sample = self.configspace.sample_configuration().get_dictionary()
+        logger.debug('done sampling a new configuration.')
+        sample['STEPS'] = budget
+        return sample
+
+    def impute_conditional_data(self, array):
+        return_array = np.empty_like(array)
+        for i in range(array.shape[0]):
+            datum = np.copy(array[i])
+            nan_indices = np.argwhere(np.isnan(datum)).flatten()
+            while(np.any(nan_indices)):
+                nan_idx = nan_indices[0]
+                valid_indices = np.argwhere(np.isfinite(array[:,nan_idx])).flatten()
+                if len(valid_indices) > 0:
+                    # pick one of them at random and overwrite all NaN values
+                    row_idx = np.random.choice(valid_indices)
+                    datum[nan_indices] = array[row_idx, nan_indices]
+                else:
+                    # no good point in the data has this value activated, so fill it with a valid but random value
+                    t = self.vartypes[nan_idx]
+                    if t == 0:
+                        datum[nan_idx] = np.random.rand()
+                    else:
+                        datum[nan_idx] = np.random.randint(t)
+                nan_indices = np.argwhere(np.isnan(datum)).flatten()
+            return_array[i,:] = datum
+        return(return_array)
+
+    def new_result(self, loss, budget, parameters, update_model=True):
+        """function to register finished runs
+
+        Every time a run has finished, this function should be called
```

Comment:
please add the docstring for parameters.

Reply:
Done